[MRG] create functional annotation lesson (#5)
mblstamps/stamps2022: functional_annotation/functional_annotation.md

         -c:  Closed ends.  Do not allow genes to run off edges.
         -d:  Write nucleotide sequences of genes to the selected file.
         -f:  Select output format (gbk, gff, or sco).  Default is gbk.
         -g:  Specify a translation table to use (default 11).
         -h:  Print help menu and exit.
         -i:  Specify FASTA/Genbank input file (default reads from stdin).
         -m:  Treat runs of N as masked sequence; don't build genes across them.
         -n:  Bypass Shine-Dalgarno trainer and force a full motif scan.
         -o:  Specify output file (default writes to stdout).
         -p:  Select procedure (single or meta).  Default is single.
         -q:  Run quietly (suppress normal stderr output).
         -s:  Write all potential genes (with scores) to the selected file.
         -t:  Write a training file (if none exists); otherwise, read and use
              the specified training file.
         -v:  Print version number and exit.
```

There are a few flags that are useful here.
First, we see `-a`, which will allow us to write out amino acid sequences (translated) ORFs.
`-d` allows us to also save the nucleotide sequences.

> Lastly, `-p` let's us tell prodigal that we're working with a metagenome where we might have fragmented contigs and partial ORFs.

Author: ```suggestion
Lastly, `-p` let's us tell prodigal that we're working with a metagenome.
By default, prodigal assumes we are working with a single genome. 
In this mode, it will study the sequences we give it to try to better learn how to predict ORFs for that genome. 
By telling it we are working with a metagenome (which holds sequences from many different genomes) it will use pre-calculated training files to guide its predictions.
```

```
exec_annotation --help
```

`kofamscan` needs a few database files in order to run.
Let's download them.

```
wget ftp://ftp.genome.jp/pub/db/kofam/ko_list.gz      # download the ko list 
wget ftp://ftp.genome.jp/pub/db/kofam/profiles.tar.gz # download the hmm profiles
```

The first file is the KEGG orthlog (KO) list.
The second file is the HMM profiles for the KOs.

We also have to unzip and untar the relevant files:
```
gunzip ko_list.gz
tar xf profiles.tar.gz
```

kofamscan runs using a config file. 
Using nano, build a config file that looks like this:

```
nano config.yaml
```

```
# Path to your KO-HMM database
# A database can be a .hmm file, a .hal file or a directory in which
# .hmm files are. Omit the extension if it is .hal or .hmm file
profile: ./profiles

# Path to the KO list file
ko_list: ko_list

# Path to an executable file of hmmsearch
# You do not have to set this if it is in your $PATH
# hmmsearch: /usr/local/bin/hmmsearch

# Path to an executable file of GNU parallel
# You do not have to set this if it is in your $PATH
# parallel: /usr/local/bin/parallel

# Number of hmmsearch processes to be run parallelly
cpu: 8
```

> 

Reviewer: ```suggestion

> You might notice the `$PATH` variable above. We didn't get to cover what this special variable is in a Unix-like environment, but it can be useful to know about. The PATH variable holds all the locations (directories) our command line will automatically search when looking for a program we try to run. If interested in more about this and how to modify our PATH, you can find more about it [here](https://astrobiomike.github.io/unix/modifying_your_path) 👍 

```

Author: ```suggestion
You might notice the `$PATH` variable above. We didn't get to cover what this special variable is in a Unix-like environment, but it can be useful to know about. The PATH variable holds all the locations (directories) our command line will automatically search when looking for a program we try to run. If interested in more about this and how to modify our PATH, you can find more about it [here](https://cac-staff.github.io/swc-bash-lesson/j05-hpc.html) 
```

```

Look at the results:

```
less SRR8859675_final_contigs_kofamscan.txt 
```

## Visualizing the output

```
mamba create -n keggdecoder python=3.6 pip
conda activate keggdecoder
pip install KEGGDecoder==1.2.2
```

```
KEGG-decoder -i SRR8859675_final_contigs_kofamscan.txt -o kegg_decoder_out --vizoption static
```

When this is done running, you'll have a file named `kegg_decoder_out.svg`. 
You can open this in jupyter notebook or transfer it to your local computer using a program like `scp`.
We've recreated the visual below so you can look at it!

![](https://i.imgur.com/kB5CuIK.png)

**Challenge**
How would this visualization change if we had run `kofamscan` on MAGS vs. on the metagenome assembly?

<details>
  <summary>Challenge Answer</summary>
    Assuming every contig was binned, we would still see all of the functions represented in the plot, but we would have a row for each MAG. The functions would be spread out across the organisms.
</details>

## Next steps/future directions

We annotated a thing! Now what?
The next steps in any workflow will depend on the goal of your study.

> If you have multiple samples, you could do ortholog or pathway level differential abundance analysis with a tool like corncob. 

Reviewer: I probably wouldn't include this line unless wanting to spend time explaining the difference between what this would be doing (differential abundance of recovered copies) vs getting coverage of these annotations and then using the summed coverage values of each annotation to look at differential abundance of the functional annotations we recovered/assigned. If you follow what i mean

Author: removed!

    Three. Every other line is commented out.
</details>

Let's run the actual annotation command now.
```
exec_annotation -f mapper --config config.yaml -o SRR8859675_final_contigs_kofamscan.txt SRR8859675_final_contigs.faa
```

Look at the results:

```
less SRR8859675_final_contigs_kofamscan.txt 
```

## Visualizing the output

```
mamba create -n keggdecoder python=3.6 pip
conda activate keggdecoder
pip install KEGGDecoder==1.2.2
```

```
KEGG-decoder -i SRR8859675_final_contigs_kofamscan.txt -o kegg_decoder_out --vizoption static
```

When this is done running, you'll have a file named `kegg_decoder_out.svg`. 
You can open this in jupyter notebook or transfer it to your local computer using a program like `scp`.
We've recreated the visual below so you can look at it!

![](https://i.imgur.com/kB5CuIK.png)

**Challenge**
How would this visualization change if we had run `kofamscan` on MAGS vs. on the metagenome assembly?

<details>
  <summary>Challenge Answer</summary>
    Assuming every contig was binned, we would still see all of the functions represented in the plot, but we would have a row for each MAG. The functions would be spread out across the organisms.
</details>

> ## Next steps/future directions

Author: ```suggestion
## Other annotation tools
```

> We annotated a thing! Now what?

Author: ```suggestion
```

> The next steps in any workflow will depend on the goal of your study.

Author: ```suggestion
```

> If you have multiple samples, you could do ortholog or pathway level differential abundance analysis with a tool like corncob. 

Author: ```suggestion
```

> If you're interested in what functional content assembles but doesn't bin, you could annotate your MAGs using the above pipeline and compare the content to the assembly.

Author: ```suggestion
```
